lib/Tools/circt-verilog-lsp-server/VerilogServerImpl/VerilogServer.cpp

```diff
@@ -0,0 +1,353 @@
+#include "slang/diagnostics/DiagnosticClient.h"
+#include "slang/diagnostics/Diagnostics.h"
+#include "slang/driver/Driver.h"
+#include "slang/syntax/SyntaxTree.h"
+#include "slang/text/SourceLocation.h"
+#include "slang/text/SourceManager.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SourceMgr.h"
+
+#include <memory>
+#include <optional>
+
+using namespace mlir;
+using namespace mlir::lsp;
+
+using namespace circt::lsp;
+using namespace circt;
+
+static mlir::lsp::DiagnosticSeverity
+getSeverity(slang::DiagnosticSeverity severity) {
+  switch (severity) {
+  case slang::DiagnosticSeverity::Fatal:
+  case slang::DiagnosticSeverity::Error:
+    return mlir::lsp::DiagnosticSeverity::Error;
+  case slang::DiagnosticSeverity::Warning:
+    return mlir::lsp::DiagnosticSeverity::Warning;
+  case slang::DiagnosticSeverity::Ignored:
+  case slang::DiagnosticSeverity::Note:
+    return mlir::lsp::DiagnosticSeverity::Information;
+  }
+  llvm_unreachable("all slang diagnostic severities should be handled");
+  return mlir::lsp::DiagnosticSeverity::Error;
+}
+namespace {
+
+// A global context carried around by the server.
+struct VerilogServerContext {
+  VerilogServerContext(const VerilogServerOptions &options)
+      : options(options) {}
+  const VerilogServerOptions &options;
+};
+
+//===----------------------------------------------------------------------===//
+// VerilogDocument
+//===----------------------------------------------------------------------===//
+
+/// This class represents all of the information pertaining to a specific
+/// Verilog document.
+class LSPDiagnosticClient;
+struct VerilogDocument {
+  VerilogDocument(VerilogServerContext &globalContext,
+                  const mlir::lsp::URIForFile &uri, StringRef contents,
+                  std::vector<mlir::lsp::Diagnostic> &diagnostics);
+  VerilogDocument(const VerilogDocument &) = delete;
+  VerilogDocument &operator=(const VerilogDocument &) = delete;
+
+  const mlir::lsp::URIForFile &getURI() const { return uri; }
+
+  llvm::SourceMgr &getSourceMgr() { return sourceMgr; }
+  llvm::SmallDenseMap<uint32_t, uint32_t> &getBufferIDMap() {
+    return bufferIDMap;
+  }
+
+  const slang::SourceManager &getSlangSourceManager() const {
+    return driver.sourceManager;
+  }
+
+  // Return LSP location from slang location.
+  mlir::lsp::Location getLspLocation(slang::SourceLocation loc) const;
+
+private:
+  // A map from slang buffer ID to the corresponding buffer ID in the LLVM
+  // source manager.
+  llvm::SmallDenseMap<uint32_t, uint32_t> bufferIDMap;
+
+  // The compilation result.
+  FailureOr<std::unique_ptr<slang::ast::Compilation>> compilation;
+
+  // The slang driver.
+  slang::driver::Driver driver;
+
+  // The LLVM source manager.
+  llvm::SourceMgr sourceMgr;
+
+  // The URI of the document.
+  const mlir::lsp::URIForFile &uri;
```

Comment:
URI might be deallocated so it's necessary to copy here.
```suggestion
  mlir::lsp::URIForFile uri;
```